Change Ollama AddModel to create a separate resource

Mrxx99 opened the pull request:
<!-- 🚨 Please Do Not skip any instructions and information mentioned below as they are all required and essential to evaluate and test the PR. By fulfilling all the required information you will be able to reduce the volume of questions and most likely help merge the PR faster 🚨 -->

<!-- ⚠️ We will not merge the PR to the main repo if your changes are not in a *feature branch* of your forked repository. Create a new branch in your repo, **do not use the `main` branch in your repo**! -->

<!-- ⚠️ **Every PR** needs to have a linked issue and have previously been approved. PRs that don't follow this will be rejected. >

<!-- 👉 It is imperative to resolve ONE ISSUE PER PR and avoid making multiple changes unless the changes interrelate with each other -->

<!-- 📝 Please always keep the "☑️ Allow edits by maintainers" button checked in the Pull Request Template as it increases collaboration with the Toolkit maintainers by permitting commits to your PR branch (only) created from your fork. This can let us quickly make fixes for minor typos or forgotten StyleCop issues during review without needing to wait on you doing extra work. Let us help you help us! 🎉 -->

**Closes #119**

<!-- Add an overview of the changes here -->
As discussed in #119 this changes the AddModel methods on Ollama resource to create and return a new OllamaModelResosurce, that can be used with WithReference instead of using the Ollama resource.
The advantage is, that now one Ollama resource can have multiple models, that are referenced by same or different applications. Previously only the default model could be used without manually setting complicated env variables.

<!-- All details should be in the linked issue. Feel free to call out any outstanding differences here. -->

## PR Checklist

<!-- Please check if your PR fulfills the following requirements, and remove the ones that are not applicable to the current PR -->

- [x] Created a feature/dev branch in your fork (vs. submitting directly from a commit on main)
- [x] Based off latest main branch of toolkit
- [x] PR doesn't include merge commits (always rebase on top of our main, if needed)
- [ ] New integration
  - [ ] Docs are written
  - [ ] Added description of major feature to project description for NuGet package (4000 total character limit, so don't push entire description over that)
- [ ] Tests for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Contains **NO** breaking changes
- [ ] Every new API (including internal ones) has full XML docs
- [ ] Code follows all style conventions

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below.
Please note that breaking changes are likely to be rejected within minor release cycles or held until major versions. -->

This is a breaking change so I guess v9 would be a good target for this.
It breaks how AddModel works previously it would return the Ollama resource back, so whenever this was used, the code would have to change to call AddModel on a variable that was returned from AddOllama instead. The WithDefaultValue still works, this limits a bit the breaking change.

## Other information

<!-- Please add any other information that might be helpful to reviewers. -->
This is still work in progress, but I already created the PR, so some design decisions can be discussed.
E.g. should it still work with a custom WithReference extension and environment variables or could a connection string that includes the model name be used instead? The PR implementation provides both but uses the custom WithReference one, if that method was removed it would use the connection string one.

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.Ollama/OllamaModelResource.cs:
```diff
@@ -0,0 +1,28 @@
+    /// <summary>
+    /// Gets the connection string expression for the Ollama model.
+    /// </summary>
+    public ReferenceExpression ConnectionStringExpression => ReferenceExpression.Create($"{Parent}?Model={ModelName}");
```
This is getting me to rethink the "connection string" concept that we have and whether we should go to a "real" connection string like `Endpoint=...;Model=...` structure.

That would be a breaking change but it does align better with how other Aspire integrations work (and we can use the `DbConnectionStringBuilder` class to parse it).

aaronpowell reviewed (commented):
This looks **amazing**, thanks @Mrxx99!

I have one question about the connection string for models and I think doing it would help address #107

aaronpowell commented:
Let's also change this to merge into the aspire-9 branch not main

Mrxx99 commented:
@aaronpowell  I did the change to use a proper DB connection string and removed the environments variables. Should the 'WithDefaultModel' approach be kept or also removed?

aaronpowell commented:
Let's use the connection string approach the whole way down. We can accept a breaking change going into the v9 release (which this will land) as we're also really only soft launched at the moment so usage is low (and thus if we were to change things "radically" it won't be a huge problem).

Same goes for the `WithDefaultModel`, let's remove that as there's now a better approach to providing that through the work you've done.

aaronpowell commented on examples/ollama/CommunityToolkit.Aspire.Hosting.Ollama.AppHost/Program.cs:
```diff
@@ -1,13 +1,14 @@
 var ollama = builder.AddOllama("ollama")
-    .AddModel("phi3")
-    .WithDefaultModel("phi3")
+    .WithDefaultModel("phi:chat")
```
As mentioned in a recent comment, let's remove the `WithDefaultModel` method and instead you always use `WithModel` and provide the model reference.

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.Ollama/OllamaResource.cs:
```diff
@@ -35,10 +35,21 @@ public class OllamaResource(string name) : ContainerResource(name), IResourceWit
+        builder.AppendLiteral($"Endpoint=");
+        builder.Append($"{PrimaryEndpoint.Property(EndpointProperty.Scheme)}://{PrimaryEndpoint.Property(EndpointProperty.Host)}:{PrimaryEndpoint.Property(EndpointProperty.Port)}");
+
+        if (includeDefaultModel && !string.IsNullOrEmpty(DefaultModel))
+        {
+            builder.AppendLiteral($";Model={DefaultModel}");
+        }
```
We'll be able to drop this with the removal of the concept of default model

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.Ollama/OllamaResourceBuilderExtensions.cs:
```diff
@@ -55,63 +51,76 @@ public static IResourceBuilder<OllamaResource> WithDataVolume(this IResourceBuil
-    /// This method adds the connection string and model references to the application model.
-    /// </remarks>
-    public static IResourceBuilder<T> WithReference<T>(this IResourceBuilder<T> builder, IResourceBuilder<OllamaResource> ollama) where T : IResourceWithEnvironment
+    public static IResourceBuilder<OllamaModelResource> AddHuggingFaceModel(this IResourceBuilder<OllamaResource> builder, string name, string modelName)
```
This method could do a pass-through to the `AddModel` with the prefix of `hf.co/`, which would reduce a little bit of code duplication.

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.Ollama/OllamaResourceBuilderExtensions.cs:
```diff
@@ -55,63 +51,76 @@ public static IResourceBuilder<OllamaResource> WithDataVolume(this IResourceBuil
+    {
+        ArgumentNullException.ThrowIfNull(builder, nameof(builder));
+        ArgumentException.ThrowIfNullOrWhiteSpace(modelName, nameof(modelName));
 
-            for (int i = 0; i < ollama.Resource.Models.Count; i++)
-            {
-                var model = ollama.Resource.Models[i];
-                context.EnvironmentVariables[$"Aspire__OllamaSharp__{resource.Name}__Models__{i}"] = model;
+        builder.AddModel(modelName);
 
-                if (model == ollama.Resource.DefaultModel)
-                {
-                    context.EnvironmentVariables[$"Aspire__OllamaSharp__{resource.Name}__SelectedModel"] = model;
-                }
-            }
-        });
+        builder.Resource.SetDefaultModel(modelName);
+        return builder;
```
As discussed, let's remove this method and instead you'll use the `OllamaModelResource` as the thing you pass in if you want a default model specified.

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.Ollama/PublicAPI.Shipped.txt:
```diff
@@ -11,10 +11,8 @@ Aspire.Hosting.ApplicationModel.OpenWebUIResource.ConnectionStringExpression.get
 Aspire.Hosting.ApplicationModel.OpenWebUIResource.OpenWebUIResource(string! name) -> void
 Aspire.Hosting.ApplicationModel.OpenWebUIResource.PrimaryEndpoint.get -> Aspire.Hosting.ApplicationModel.EndpointReference!
 Aspire.Hosting.OllamaResourceBuilderExtensions
-static Aspire.Hosting.OllamaResourceBuilderExtensions.AddModel(this Aspire.Hosting.ApplicationModel.IResourceBuilder<Aspire.Hosting.ApplicationModel.OllamaResource!>! builder, string! modelName) -> Aspire.Hosting.ApplicationModel.IResourceBuilder<Aspire.Hosting.ApplicationModel.OllamaResource!>!
```
We need to leave this method here and instead it'll get marked with REMOVED in the unshipped file.

@Alirexaa we should document this in the contributing guide better.

aaronpowell commented on src/CommunityToolkit.Aspire.Hosting.Ollama/PublicAPI.Shipped.txt:
```diff
@@ -11,10 +11,8 @@ Aspire.Hosting.ApplicationModel.OpenWebUIResource.ConnectionStringExpression.get
 static Aspire.Hosting.OllamaResourceBuilderExtensions.WithDataVolume(this Aspire.Hosting.ApplicationModel.IResourceBuilder<Aspire.Hosting.ApplicationModel.OllamaResource!>! builder, string? name = null, bool isReadOnly = false) -> Aspire.Hosting.ApplicationModel.IResourceBuilder<Aspire.Hosting.ApplicationModel.OllamaResource!>!
 static Aspire.Hosting.OllamaResourceBuilderExtensions.WithDefaultModel(this Aspire.Hosting.ApplicationModel.IResourceBuilder<Aspire.Hosting.ApplicationModel.OllamaResource!>! builder, string! modelName) -> Aspire.Hosting.ApplicationModel.IResourceBuilder<Aspire.Hosting.ApplicationModel.OllamaResource!>!
 static Aspire.Hosting.OllamaResourceBuilderExtensions.WithOpenWebUI<T>(this Aspire.Hosting.ApplicationModel.IResourceBuilder<T!>! builder, System.Action<Aspire.Hosting.ApplicationModel.IResourceBuilder<Aspire.Hosting.ApplicationModel.OpenWebUIResource!>!>? configureContainer = null, string? containerName = null) -> Aspire.Hosting.ApplicationModel.IResourceBuilder<T!>!
-static Aspire.Hosting.OllamaResourceBuilderExtensions.WithReference<T>(this Aspire.Hosting.ApplicationModel.IResourceBuilder<T>! builder, Aspire.Hosting.ApplicationModel.IResourceBuilder<Aspire.Hosting.ApplicationModel.OllamaResource!>! ollama) -> Aspire.Hosting.ApplicationModel.IResourceBuilder<T>!
```
We need to leave this method here and instead it'll get marked with REMOVED in the unshipped file.

aaronpowell commented on src/CommunityToolkit.Aspire.OllamaSharp/AspireOllamaSharpExtensions.cs:
```diff
@@ -10,7 +11,7 @@ namespace Microsoft.Extensions.Hosting;
 public static class AspireOllamaSharpExtensions
 {
-    private const string DefaultConfigSectionName = "Aspire:OllamaSharp";
+    private const string DefaultConfigSectionName = "Aspire:Ollama";
```
This needs to stick with `OllamaSharp` as that's the client this is relevant for

aaronpowell reviewed (changes requested):
Got a few more changes in here but it's shaping up really nicely.

I'm loving how the integration now looks

mitchdenny commented:
This PR is kind of setting a convention for how to pass around the model name in the connection string. I think this is probably the right thing to do but I know that this has been a contentious issue so should be discussed.

For me the most important thing is the understand how a developer moves between different API providers. So for local development they are probably going to use Ollama, but when they deploy do they want to adopt Open AI/Azure Open AI. Can they host things

```csharp
var builder = DistributedApplication.CreateBuilder(args);
IResourceBuilder<IResourceWithConnectionString> model = builder.ExecutionContext.IsRunMode
                                                          ? builder.AddOllama(...).AddModel(...)
                                                          : builder.AddAzureOpenAI(...).AddModel(...);
```

aaronpowell reviewed (approved):
I feel like this is ready to go

aaronpowell commented:
> This PR is kind of setting a convention for how to pass around the model name in the connection string. I think this is probably the right thing to do but I know that this has been a contentious issue so should be discussed.
> 
> For me the most important thing is the understand how a developer moves between different API providers. So for local development they are probably going to use Ollama, but when they deploy do they want to adopt Open AI/Azure Open AI. Can they host things
> 
> ```cs
> var builder = DistributedApplication.CreateBuilder(args);
> IResourceBuilder<IResourceWithConnectionString> model = builder.ExecutionContext.IsRunMode
>                                                           ? builder.AddOllama(...).AddModel(...)
>                                                           : builder.AddAzureOpenAI(...).AddModel(...);
> ```

There are some complexities that come with transitioning between different AI services, but I think that the introduction of the Microsoft.Extensions.AI (and the updates coming in #164 to support it for Ollama) will help with that by having the client abstraction away and then it becomes a case of the client integration <-> hosting integration, as you'd likely use OllamaSharp + Ollama Hosting then move to the OpenAI/Azure OpenAI set.

aaronpowell commented:
@Mrxx99 - I think we're looking good, anything left from your end before putting this in for a merge?

aaronpowell commented:
> Should we keep the AddModel(string) overload where the resource name is generated? I am not 100% sure about the sanitizing.

I think we can combine them into a single method, make `modelName` optional and if it's not provided we use `name` for it.

Digging through Aspire's source, I can't see anywhere that would have a problem if the resource name as non-alphanumeric characters, so we could probably get away with a resource named `phi3:chat` or `huggingface.io/something`.

aaronpowell commented:
I've decided to merge it as-is so I can get a head start on the merge conflicts that will come with #164 

Thanks again, I'm really excited at how the Ollama integration is now looking.
